app/stacksnipe/stacksnipe.go

```diff
@@ -0,0 +1,227 @@
+				names = append(names, c.Name)
+				cliParams = append(cliParams, c.CLIParams)
+			}
+
+			i.metricsFunc(names, cliParams)
+		}
+	}
+}
+
+// snipe reads /proc entries from procPath, looking for processes that look like Ethereum validator stack components.
+func snipe(ctx context.Context, procPath string) ([]StackComponent, error) {
+	var (
+		wb      = make(chan StackComponent)
+		ret     []StackComponent
+		walkErr error
+	)
+
+	go func() {
+		if err := filepath.WalkDir(procPath, walkFunc(ctx, wb)); err != nil {
+			walkErr = errors.Wrap(err, "cannot walk proc path", z.Str("proc_path", procPath))
+		}
+
+		close(wb)
+	}()
+
+	for c := range wb {
+		ret = append(ret, c)
+	}
+
+	if walkErr != nil {
+		return nil, walkErr
+	}
+
+	return ret, nil
+}
+
+// walkFunc walks a /proc-like filesystem as invoked by filepath.WalkDir, and sends entries to wb.
+func walkFunc(ctx context.Context, wb chan<- StackComponent) fs.WalkDirFunc {
+	cmdlineDedup := make(map[string]struct{})
+
+	return func(path string, d fs.DirEntry, err error) error {
+		// ignore directory access error and don't walk the directory
+		if err != nil {
+			return nil //nolint:nilerr // best effort component
+		}
+
+		// ignore files
+		if !d.IsDir() {
+			return nil
+		}
+
+		// ignore directories which don't look like pids
+		hostPID, err := strconv.ParseUint(d.Name(), 10, 64)
+		if err != nil {
+			return nil //nolint:nilerr // best effort component
+		}
+
+		// do initial filtering by process' comm
+		commBytes, err := os.ReadFile(filepath.Join(path, "comm"))
+		if err != nil {
+			// ignore error, best effort
+			return nil //nolint:nilerr // best effort component
+		}
+
+		comm := strings.TrimSpace(string(commBytes))
+		_, vcOk := supportedVCs[comm]
+		_, maybeVCOk := maybeVCs[comm]
+
+		if !vcOk && !maybeVCOk {
+			return nil
+		}
+
+		// grab vc's cmdline
+		cmdlineBytes, err := os.ReadFile(filepath.Join(path, "cmdline"))
+		if err != nil {
+			// ignore error, best effort
+			return nil //nolint:nilerr // best effort component
+		}
+
+		cmdlineString := string(cmdlineBytes)
+
+		cmdlineSplit := bytes.Split(cmdlineBytes, []byte{0})
+
+		var vcName string
+		for vc := range supportedVCs {
+			if strings.Contains(cmdlineString, vc) {
+				vcName = vc
+			}
+		}
+
+		if vcName == "" {
+			return nil
+		}
+
+		if _, ok := cmdlineDedup[cmdlineString]; ok {
+			// we already have seen this, probably a background thread
+			return nil
+		}
+
+		cmdlineDedup[cmdlineString] = struct{}{}
+
+		var cmdLine []string
+		for _, cl := range cmdlineSplit {
+			if len(cl) == 0 {
+				continue
+			}
+
+			cmdLine = append(cmdLine, string(cl))
+		}
+
+		if len(cmdLine) == 0 {
+			// no cmdline, ignore
+			return nil
+		}
+
+		cmdLineStr := strings.Join(cmdLine, " ")
+
+		log.Debug(ctx, "Detected stack component", z.Str("name", vcName), z.U64("host_pid", hostPID), z.Str("cmdline", cmdLineStr))
+
+		wb <- StackComponent{
+			Name:      vcName,
+			CLIParams: cmdLineStr,
+		}
+
+		return nil
```

Comment:
Maybe to respect `ctx` state to some extent, we could return `ctx.Err()`?

Reply:
We could, but I figured it's not a big deal in this component, we just want to stop the servicce.